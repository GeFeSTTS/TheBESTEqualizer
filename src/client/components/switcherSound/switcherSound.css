.SwitcherContainer {
    background: var(--headerBackground-dark-super);
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    width: 100%;
    max-width: 1024px;
    justify-content: flex-end;
    color: var(--authFont-white);
    -webkit-box-pack: justify;
    -ms-flex-pack: justify;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
}

.SwitcherContainer .icon {
    color: var(--mainText-accent-blue);
    margin-right: 5px;
}

.SwitcherContainer .voiceText {
    margin-left: 10px;
}


.SwitcherContainer--slider {
    margin-left: 10px;
    width: 200px;
    height: 4px;
}

.SwitcherContainer--slider .rangeslider__fill {
    background-color: var(--rangeSlider-ocean);
}

.SwitcherContainer .rangeslider__handle {
    width: 17px;
    height: 17px;
    border-radius: 50%;
    outline: none;
    border: none;
    background-color: var(--mainText-accent-blue);
}

.SwitcherContainer--slider .rangeslider__handle .rangeslider__handle-tooltip {
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    -webkit-touch-callout: none;
}

.SwitcherContainer--slider .rangeslider__handle:after {
    display: none;
}

@media only screen and (max-width: 1020px) {
    .SwitcherContainer--slider {
        width: 100px;
    }
}

@media only screen and (max-width: 720px) {
    .SwitcherContainer {
        justify-content: center;
    }
}

