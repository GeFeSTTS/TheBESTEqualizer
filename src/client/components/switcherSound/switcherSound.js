import React, { Component, Fragment } from 'react';
import 'react-rangeslider/lib/index.css';
import './switcherSound.css';
import Slider from 'react-rangeslider';
import { connect } from 'react-redux';

export class SwitcherSound extends Component {
  state = {
    volumeValueTrack: 0.5,
    volumeValueVoice: 0.5,
    track: null,
    voice: null,
  };

  static getDerivedStateFromProps(props, state) {
    const { audioData: { sound, voice } } = props;
    const { volumeValueTrack, volumeValueVoice } = state;
    if (sound && voice) {
      sound.volume = volumeValueTrack;
      voice.volume = volumeValueVoice;
      return { track: sound, voice };
    }
    return null;
  }

  changeTracksVolume = () => {
    const {
      volumeValueTrack, volumeValueVoice, track, voice,
    } = this.state;
    const switcherTrackSound = 1 - volumeValueTrack;
    const switcherVoiceSound = 1 - volumeValueVoice;
    track.volume = switcherTrackSound;
    voice.volume = switcherVoiceSound;
  }

  changeVolume = (volumeValueTrack) => {
    const revertVolumeValueTrack = 1 - volumeValueTrack;
    this.setState({
      volumeValueTrack: parseFloat(volumeValueTrack.toFixed(2)),
      volumeValueVoice: parseFloat(revertVolumeValueTrack.toFixed(2)),
    }, () => this.changeTracksVolume());
  }

  render() {
    const { volumeValueTrack } = this.state;
    const minSliderVolume = 0;
    const maxSliderVolume = 1;
    const stepSliderVolume = 0.001;
    return (
      <div className="SwitcherContainer">
        <Fragment>
          <span>Track</span>
          <Slider
            className="SwitcherContainer--slider"
            value={volumeValueTrack}
            min={minSliderVolume}
            max={maxSliderVolume}
            step={stepSliderVolume}
            onChange={this.changeVolume}
          />
          <span className="voiceText">Voice</span>
        </Fragment>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  audioData: state.audioData,
});

export default connect(mapStateToProps)(SwitcherSound);
