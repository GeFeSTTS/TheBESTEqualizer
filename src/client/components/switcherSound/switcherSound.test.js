import React from 'react';
import { shallow } from 'enzyme';
import { SwitcherSound } from './switcherSound';

describe('TEST SOUND SWITCHER', () => {
  let switcherSoundComponent;
  const props = {
    min: 0,
    max: 1,
    step: 0.001,
    onChange: () => {},
    audioData: { sound: {}, voice: {} },
  };

  beforeEach(() => {
    switcherSoundComponent = shallow(<SwitcherSound {...props} />);
  });

  it('is render component', () => {
    switcherSoundComponent.setState({ volumeValueTrack: 0 });
    expect(switcherSoundComponent).toHaveLength(1);
  });

  it('render component correctly', () => {
    switcherSoundComponent.setState({ volumeValueTrack: 0 });
    switcherSoundComponent.find('Slider');
    expect(switcherSoundComponent).toMatchSnapshot();
  });

  it('should call function after change', () => {
    switcherSoundComponent.setState({ volumeValueTrack: 0.8 });
    const instance = switcherSoundComponent.instance();

    const spySwitching = jest.spyOn(instance, 'changeVolume');
    switcherSoundComponent.instance().forceUpdate();
    switcherSoundComponent.find('Slider').simulate('change', 1);
    expect(spySwitching).toHaveBeenCalled();
  });

  it('should we return null in getDerivedStateFromProps', () => {
    switcherSoundComponent.setProps({ audioData: { sound: null, voice: {} } });
    switcherSoundComponent.setState({ volumeValueTrack: 0 });
    expect(switcherSoundComponent).toHaveLength(1);
  });
});
