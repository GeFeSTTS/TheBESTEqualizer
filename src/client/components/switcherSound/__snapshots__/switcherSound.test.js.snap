// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST SOUND SWITCHER render component correctly 1`] = `ShallowWrapper {}`;
